File: graph/WeightedUndirectedGraph.ts
import * as readline from "readline";

class Pair {
  constructor(public first: number, public second: number) {}

  compareTo(other: Pair): number {
    if (this.first !== other.first) return this.first - other.first;
    return this.second - other.second;
  }

  toString(): string {
    return `${this.first} ${this.second}`;
  }
}

class Triplet {
  constructor(
    public first: number,
    public second: number,
    public third: number
  ) {}

  compareTo(other: Triplet): number {
    if (this.first !== other.first) return this.first - other.first;
    if (this.second !== other.second) return this.second - other.second;
    return this.third - other.third;
  }

  toString(): string {
    return `${this.first} ${this.second} ${this.third}`;
  }
}

class WeightedUndirectedGraph {
  adjList: Pair[][];
  edgeList: Triplet[];

  constructor(n: number) {
    this.adjList = Array.from({ length: n + 1 }, () => []);
    this.edgeList = [];
  }

  addEdge(u: number, v: number, w: number): void {
    this.edgeList.push(new Triplet(u, v, w));
    this.adjList[u].push(new Pair(v, w));
    this.adjList[v].push(new Pair(u, w));
  }

  printEdgeList(): void {
    console.log("Edge List:");
    console.log(this.edgeList.map((e) => e.toString()).join("\n"));
  }

  printAdjList(): void {
    console.log("Adjacency List:");
    for (let i = 0; i < this.adjList.length; i++) {
      const neighbors = this.adjList[i];
      if (neighbors.length === 0) {
        console.log(`${i}: No neighbors`);
      } else {
        const neighborStr = neighbors.map((p) => `(${p})`).join(" ");
        console.log(`${i}: ${neighborStr}`);
      }
    }
  }
}

async function main(): Promise<void> {
  const rl = readline.createInterface({
    input: process.stdin,
    terminal: false,
  });

  const lines: string[] = [];

  for await (const line of rl) {
    lines.push(line.trim());
  }

  let idx = 0;
  const tokens: string[] = lines.flatMap((line) => line.split(/\s+/));

  const readInt = (): number => parseInt(tokens[idx++], 10);

  let t = readInt();

  while (t-- > 0) {
    const V = readInt();
    const E = readInt();

    const g = new WeightedUndirectedGraph(V);

    for (let i = 0; i < E; i++) {
      const u = readInt();
      const v = readInt();
      const w = readInt();
      g.addEdge(u, v, w);
    }

    // SOLUTION LOGIC HERE

    g.printEdgeList();
    g.printAdjList();
  }
}

main();

File: graph/WeightedUndirectedGraph.py
import sys
from sys import stdin

class Pair:
    def __init__(self, first, second):
        self.first = first
        self.second = second

    def __lt__(self, other):
        if self.first != other.first:
            return self.first < other.first
        return self.second < other.second

    def __eq__(self, other):
        return self.first == other.first and self.second == other.second

    def __repr__(self):
        return f"{self.first} {self.second}"


class Triplet:
    def __init__(self, first, second, third):
        self.first = first
        self.second = second
        self.third = third

    def __lt__(self, other):
        if self.first != other.first:
            return self.first < other.first
        if self.second != other.second:
            return self.second < other.second
        return self.third < other.third

    def __eq__(self, other):
        return (self.first == other.first and
                self.second == other.second and
                self.third == other.third)

    def __repr__(self):
        return f"{self.first} {self.second} {self.third}"


class WeightedUndirectedGraph:
    def __init__(self, n):
        self.adj_list = [[] for _ in range(n + 1)]
        self.edge_list = []

    def add_edge(self, u, v, w):
        self.edge_list.append(Triplet(u, v, w))
        self.adj_list[u].append(Pair(v, w))
        self.adj_list[v].append(Pair(u, w))

    def print_edge_list(self):
        print("Edge List:")
        print("\n".join(str(e) for e in self.edge_list))

    def print_adj_list(self):
        print("Adjacency List:")
        for i, neighbors in enumerate(self.adj_list):
            if not neighbors:
                print(f"{i}: No neighbors")
            else:
                neighbor_str = " ".join(f"({p})" for p in neighbors)
                print(f"{i}: {neighbor_str}")


def main():
    input = stdin.read().split()
    idx = 0

    t = int(input[idx]); idx += 1

    while t:
        t -= 1
        V = int(input[idx]); idx += 1
        E = int(input[idx]); idx += 1

        g = WeightedUndirectedGraph(V)

        for _ in range(E):
            u = int(input[idx]); idx += 1
            v = int(input[idx]); idx += 1
            w = int(input[idx]); idx += 1
            g.add_edge(u, v, w)

        # SOLUTION LOGIC HERE

        g.print_edge_list()
        g.print_adj_list()


if __name__ == "__main__":
    main()
